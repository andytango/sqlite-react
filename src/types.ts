import { EventEmitter } from "@andytango/ts-event-emitter";

export type DbEventEmitter = EventEmitter<DbEventMap>;

export interface DbOpts {
  sqlDataUrl: string;
  sqlJsWorkerPath: string;
  worker?: Worker;
  getDbFile?: (s: string) => Promise<ArrayBuffer>;
}

export interface DbManager extends DbEventEmitter {
  init: () => Promise<void>;
  exec: DbExec;
  terminate: DbWorker["terminate"];
}

export interface DbResponse {
  error?: string;
  results?: DbResult[];
}

export interface DbResult {
  columns: string[];
  values: [][];
}

export type DbExec = (sql: string) => Promise<DbResponse>;

export type DbWorker = Worker;

export type DbQueryFormatter = (...a: any[]) => string;

export interface DbQueryState<T> {
  loading: boolean;
  results: T[];
}

export type SetDbQueryState<T> = React.Dispatch<
  React.SetStateAction<DbQueryState<T>>
>;

export interface DbEvent extends DbOpts {}

interface QueryEvent extends DbOpts {
  queryId: string;
  sql: string;
}

export interface QueryStartEvent extends QueryEvent {
  startedAt: number;
}

export interface QueryResultEvent extends QueryEvent {
  results: DbResult[];
  startedAt: number;
  completedAt: number;
}

export interface QueryErrorEvent extends QueryEvent {
  error: string;
  startedAt: number;
  completedAt: number;
}

export type DbEventMap = {
  dbInit: DbEvent;
  dbReady: DbEvent;
  queryStart: QueryStartEvent;
  queryResult: QueryResultEvent;
  queryError: QueryErrorEvent;
};

export interface DbContextState {
  db: DbManager;
  queries: DbQueries;
  sqlDataUrl: string;
  sqlJsWorkerPath: string;
}

export type DbQueries = Record<number, DbQueryState<unknown>>;
export interface DbQueryState<T> {
  loading: boolean;
  results: T[];
  error?: string;
  sql: string;
}

export type DbAction =
  | {
      type: "init";
      db: DbManager;
      sqlDataUrl: string;
      sqlJsWorkerPath: string;
    }
  | {
      type: "query_exec";
      queryId: number;
      sql: string;
    }
  | {
      type: "query_error";
      queryId: number;
      error: string;
    }
  | {
      type: "query_result";
      queryId: number;
      results: unknown[];
    };
