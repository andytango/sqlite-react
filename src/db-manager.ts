import { createDbEventEmitter, DbEventEmitter } from "./db-event-emitter";
import { createDbWorker, DbWorker } from "./db-worker";
import { DbExec, DbOpts, DbResponse } from "./types";

interface DbManager extends DbEventEmitter {
  init: () => Promise<void>;
  exec: DbExec;
  terminate: DbWorker["terminate"];
}

interface Context extends DbOpts {
  dbWorker: DbWorker;
  emitter: DbEventEmitter;
  dbInstance: DbExec | null;
  isDbInitialising: boolean;
  onInitialised: ((dbExec: DbExec) => void)[];
  getDbFile: (s: string) => Promise<ArrayBuffer>;
}

export function createDbManager(opts: DbOpts): DbManager {
  const context = createInitialState(opts);
  const { dbWorker, emitter } = context;
  const { terminate } = dbWorker;
  let queryIdSec = 0;

  async function init() {
    await getInstance(context);
  }

  function exec(sql: string) {
    return execQuery(context, sql, queryIdSec++);
  }

  return { ...emitter, init, exec, terminate };
}

function createInitialState(opts: DbOpts): Context {
  const dbWorker = createDbWorker(opts);
  const emitter = createDbEventEmitter();
  const { getDbFile = fetchDbFile } = opts;

  return {
    ...opts,
    dbWorker,
    emitter,
    dbInstance: null,
    isDbInitialising: false,
    onInitialised: [],
    getDbFile,
  };
}

async function fetchDbFile(url: string) {
  const res = await fetch(url);
  return res.arrayBuffer();
}

function getInstance(context: Context) {
  const { dbInstance, isDbInitialising } = context;

  if (dbInstance) {
    return Promise.resolve(dbInstance);
  }

  if (isDbInitialising) {
    return new Promise(enqueueDbInitCallback(context));
  }

  return initDb(context);
}

function enqueueDbInitCallback(context: Context) {
  return (res: (dbExec: DbExec) => void) =>
    context.onInitialised.push((dbExec) => res(dbExec));
}

async function initDb(context: Context): Promise<DbExec> {
  setDbInitialising(context);
  await loadSqliteFile(context);
  setDbReady(context);

  return setupDbExec(context);
}

function setDbInitialising(context: Context) {
  const { emitter, sqlDataUrl, sqlJsWorkerPath } = context;
  context.isDbInitialising = true;
  emitter.emit<"dbInit">("dbInit", { sqlDataUrl, sqlJsWorkerPath });
}

function execDbInitCallbacks(context: Context, exec: DbExec) {
  context.onInitialised.forEach((fn) => fn(exec));
  context.onInitialised = [];
}

async function loadSqliteFile(context: Context) {
  const { dbWorker, sqlDataUrl, getDbFile } = context;
  const sqlFile: ArrayBuffer = await getDbFile(sqlDataUrl);
  await dbWorker.open(sqlFile);
}

function setDbReady(context: Context) {
  const { emitter, sqlDataUrl, sqlJsWorkerPath } = context;
  context.isDbInitialising = false;
  emitter.emit<"dbReady">("dbReady", { sqlDataUrl, sqlJsWorkerPath });
}

function setupDbExec(context: Context) {
  const { dbWorker } = context;
  const exec = createDbExec(dbWorker);

  context.dbInstance = exec;
  execDbInitCallbacks(context, exec);

  return exec;
}

function createDbExec(worker: DbWorker) {
  return async (sql: string) => {
    const { error, results } = await worker.exec(sql);
    return { error, results };
  };
}

async function execQuery(context: Context, sql: string, queryId: number) {
  const dbExec = await getInstance(context);
  const startedAt = performance.now();

  emitQueryStartEvent(context, sql, queryId, startedAt);
  const res = await dbExec(sql);
  emitQueryCompleteEvents(context, sql, queryId, startedAt, res);

  return res;
}

function emitQueryStartEvent(
  context: Context,
  sql: string,
  queryId: number,
  startedAt: number
) {
  const { emitter, sqlDataUrl, sqlJsWorkerPath, dbWorker } = context;
  emitter.emit<"queryStart">("queryStart", {
    sqlDataUrl,
    sqlJsWorkerPath,
    queryId: queryId.toString(),
    sql,
    startedAt,
  });
}

function emitQueryCompleteEvents(
  context: Context,
  sql: string,
  queryId: number,
  startedAt: number,
  res: DbResponse
) {
  const { emitter } = context;
  const { error, results } = res;
  const base = getQueryCompleteEventBase(context, queryId, sql, startedAt);

  if (error) {
    emitter.emit<"queryError">("queryError", { ...base, error });
  } else if (results) {
    emitter.emit<"queryResult">("queryResult", { ...base, results });
  }
}
function getQueryCompleteEventBase(
  context: Context,
  queryId: number,
  sql: string,
  startedAt: number
) {
  const { sqlDataUrl, sqlJsWorkerPath } = context;
  return {
    sqlDataUrl,
    sqlJsWorkerPath,
    queryId: queryId.toString(),
    sql,
    startedAt,
    completedAt: performance.now(),
  };
}
