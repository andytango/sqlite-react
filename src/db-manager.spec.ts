import "jest-extended";
import { DbOpts } from "./types";
import { createDbManager } from "./db-manager";
import { readFileSync } from "fs";

const opts: DbOpts = {
  sqlDataUrl: "src/test-db.sqlite",
  sqlJsWorkerPath: "src/test-worker.js",
  getDbFile,
};

async function getDbFile() {
  return new ArrayBuffer(0);
}

describe("createDbManager", () => {
  it("dispatches db init events", async () => {
    const dbManager = createDbManager(opts);
    const fn1 = jest.fn();
    const fn2 = jest.fn();

    dbManager.once("dbInit", fn1);
    dbManager.once("dbReady", fn2);
    await dbManager.getInstance();

    expect(fn1).toHaveBeenCalled();
    expect(fn2).toHaveBeenCalled();

    expect(fn1).toHaveBeenCalledBefore(fn2);

    dbManager.terminate();
  });

  it("initialises and resolves first promise when ready", async () => {
    const dbManager = createDbManager(opts);
    const fn1 = jest.fn();
    const fn2 = jest.fn();
    const fn3 = jest.fn();

    dbManager.once("dbInit", fn1);
    dbManager.once("dbReady", fn2);
    await dbManager.getInstance();
    fn3();

    expect(fn1).toHaveBeenCalledBefore(fn3);
    expect(fn2).toHaveBeenCalledBefore(fn3);

    dbManager.terminate();
  });

  it("enqueues promise resolutions for all callbacks", async () => {
    const dbManager = createDbManager(opts);
    const fn1 = jest.fn();
    const fn2 = jest.fn();
    const fn3 = jest.fn();
    const fn4 = jest.fn();

    dbManager.once("dbInit", fn1);
    dbManager.once("dbReady", fn2);

    await Promise.all([
      dbManager.getInstance(),
      dbManager.getInstance().then(fn3),
      dbManager.getInstance().then(fn4),
    ]);

    expect(fn1).toHaveBeenCalledBefore(fn2);
    expect(fn2).toHaveBeenCalledBefore(fn3);
    expect(fn3).toHaveBeenCalledBefore(fn4);

    dbManager.terminate();
  });

  it("execs queries", async () => {
    const dbManager = createDbManager(opts);
    const db = await dbManager.getInstance();
    const res = await db(`select 1 as val`);

    expect(res).toMatchSnapshot();

    dbManager.terminate();
  });

  it("loads the supplied database file url", async () => {
    const dbManager = createDbManager(opts);
    const db = await dbManager.getInstance();
    const res = await db(`select example_col from example_table`);

    expect(res).toMatchSnapshot();

    dbManager.terminate();
  });
});
