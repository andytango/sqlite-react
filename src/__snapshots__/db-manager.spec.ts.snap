// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDbManager execs queries 1`] = `
Object {
  "error": undefined,
  "results": Array [
    Object {
      "columns": Array [
        "val",
      ],
      "values": Array [
        Array [
          1,
        ],
      ],
    },
  ],
}
`;

exports[`createDbManager loads the supplied database file url 1`] = `
Object {
  "error": undefined,
  "results": Array [
    Object {
      "columns": Array [
        "example_col",
      ],
      "values": Array [
        Array [
          "example_val",
        ],
      ],
    },
  ],
}
`;
